README.md
# Pareen
[![Docs Status](https://docs.rs/pareen/badge.svg)](https://docs.rs/pareen)
[![license](http://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/leod/pareen/blob/master/LICENSE)
[![Crates.io](https://img.shields.io/crates/v/pareen.svg)](https://crates.io/crates/pareen)

Pareen is a small Rust library for *par*ameterized inbetw*een*ing.
The intended application is in game programming, where you sometimes have 
two discrete game states between which you want to transition smoothly for
visualization purposes.

Pareen gives you tools for composing animations that are parameterized by
time (i.e. mappings from time to some animated value) without constantly
having to pass around time variables; it hides the plumbing, so that you
need to provide time only once: when evaluating the animation.

Animations are composed similarly to Rust's iterators, so no memory
allocations are necessary. The downside to this is that it is diffcult to store
pareen's animations. The recommended approach is to construct and evaluate
animations on the fly.

## Usage
Just add this line to your dependencies in `Cargo.toml`:
```
pareen = "0.2"
```

## Current Status
I consider `pareen` to be an experimental approach, and I'm not sure if I'm
still happy with it.  Anyway, the integration of easing functions could use
some love. Contributions are very much welcome!

Unfortunately, it looks like heavily nested animations can cause an exponential
slowdown in compilation time. For now, boxing intermediate animations serves as
a workaround, but this comes with a decrease of both readability and efficiency.

## Examples
```rust
// An animation returning a constant value
let anim1 = pareen::constant(1.0f64);

// Animations can be evaluated at any time
let value = anim1.eval(0.5);

// Animations can be played in sequence
let anim2 = anim1.seq(0.7, pareen::prop(0.25) + 0.5);

// Animations can be composed and transformed in various ways
let anim3 = anim2
    .lerp(pareen::circle().cos())
    .scale_min_max(5.0, 10.0)
    .backwards(1.0)
    .squeeze(3.0, 0.5..=1.0);

let anim4 = pareen::cubic(&[1.0, 2.0, 3.0, 4.0]) - anim3;

let value = anim4.eval(1.0);
```

### Easer
Optionally, you can enable support for integrating easing functions from
[`easer`](https://docs.rs/easer/0.2.1/easer/index.html). 

```rust
let first_anim = pareen::constant(2.0);
let second_anim = pareen::prop(1.0f32);

// Transition from first_anim to second_anim at time 0.5, applying cubic easing
// for 0.3 time units.
let anim = first_anim.seq_ease_in_out(
    0.5,
    easer::functions::Cubic,
    0.3,
    second_anim,
);
```

### Plots
There is an example that shows some animations as plots via
[RustGnuplot](https://github.com/SiegeLord/RustGnuplot) in
[examples/plots.rs](examples/plots.rs). Given that `gnuplot` has been
installed, it can be executed like this:
```bash
cargo run --example plots --feature easer
```

If everything works, you should see something like this:
![plots of the example/plots.rs animations](images/plots.png)


CHANGELOG.md
# Changelog
## Version 0.3.2 (2023-06-03)
- Bump version

## Version 0.3.1 (2022-08-31)
- Fix docs

## Version 0.3.0 (2022-08-31)
- Make operator overloading of `Mul`, `Add` and `Sub` more flexible.
  This may break compilation in some cases, since types are more generic now.
- Add `AnimWithDur` for easier composition of animations that have a fixed duration
- Internal refactoring: split into multiple modules
- Implement `Anim::{fst,snd,copied}`
- Implement `AnimWithDur::{sum,mean}` and simple linear regression
- Implement `Anim::{into_fn,into_box_fn}`

## Version 0.2.6 (2020-08-17)
- Make exponential slowdown of compile times less likely
- Add `cycle`
- Add `quadratic`

## Version 0.2.5 (2020-07-18)
- Add `Anim::repeat`, `Anim::hold` and `Anim::seq_continue`
- Allow boxed animations
- Expose `easer` library

## Version 0.2.4 (2020-07-16)
Yanked.

## Version 0.2.3 (2020-04-28)
- Fix compilation issue on rustc 1.43.0 (https://github.com/leod/pareen/pull/7)

## Version 0.2.1 (2020-01-18)
- Implement `Anim::as_ref`

## Version 0.2.0 (2020-01-13)
- `squeeze` no longer switches to a default value outside of the given range.
 Use `squeeze_and_surround` as a replacement.

## Version 0.1.3 (2019-12-03)
- Render README.md on crates.io

## Version 0.1.2 (2019-12-02)
- No change

## Version 0.1.1 (2019-12-02)
- Improve documentation

## Version 0.1.0 (2019-12-02)
- Initial version


src/stats.rs
use core::ops::{Add, Div, Mul};

use num_traits::{AsPrimitive, Float, Zero};

use crate::{Anim, AnimWithDur, Fun};

impl<F> AnimWithDur<F>
where
    F: Fun<T = usize>,
{
    pub fn fold<B, G>(&self, init: B, mut f: G) -> B
    where
        G: FnMut(B, F::V) -> B,
    {
        let mut b = init;

        for t in 0..self.1 {
            b = f(b, self.0.eval(t))
        }

        b
    }
}

impl<F> AnimWithDur<F>
where
    F: Fun<T = usize>,
    F::V: Add<Output = F::V> + Zero,
{
    pub fn sum(&self) -> F::V {
        self.fold(Zero::zero(), |a, b| a + b)
    }
}

impl<F> AnimWithDur<F>
where
    F: Fun<T = usize>,
    F::T: Clone,
    F::V: 'static + Add<Output = F::V> + Div<Output = F::V> + Zero + Copy,
    usize: AsPrimitive<F::V>,
{
    pub fn mean(&self) -> F::V {
        let len = self.1.clone().as_();
        self.sum() / len
    }
}

#[derive(Debug, Clone)]
pub struct Line<V> {
    pub y_intercept: V,
    pub slope: V,
}

impl<V> Fun for Line<V>
where
    V: Add<Output = V> + Mul<Output = V> + Clone,
{
    type T = V;
    type V = V;

    fn eval(&self, t: V) -> V {
        self.y_intercept.clone() + self.slope.clone() * t
    }
}

pub fn simple_linear_regression_with_slope<V, F, A>(slope: V, values: A) -> Anim<Line<V>>
where
    V: 'static + Float + Copy,
    F: Fun<T = usize, V = (V, V)>,
    A: Into<AnimWithDur<F>>,
    usize: AsPrimitive<V>,
{
    // https://en.wikipedia.org/wiki/Simple_linear_regression#Fitting_the_regression_line
    let values = values.into();
    let (x, y) = values.unzip();
    let x_mean = x.as_ref().mean();
    let y_mean = y.as_ref().mean();

    let y_intercept = y_mean - slope * x_mean;

    Anim(Line { y_intercept, slope })
}

pub fn simple_linear_regression<V, F, A>(values: A) -> Anim<Line<V>>
where
    V: 'static + Float + Copy,
    F: Fun<T = usize, V = (V, V)>,
    A: Into<AnimWithDur<F>>,
    usize: AsPrimitive<V>,
{
    // https://en.wikipedia.org/wiki/Simple_linear_regression#Fitting_the_regression_line
    let values = values.into();
    let (x, y) = values.unzip();
    let x_mean = x.as_ref().mean();
    let y_mean = y.as_ref().mean();
    let numerator = values
        .as_ref()
        .map(|(x, y)| (x - x_mean) * (y - y_mean))
        .sum();
    let denominator = x.as_ref().map(|x| (x - x_mean) * (x - x_mean)).sum();
    let slope = numerator / denominator;

    let y_intercept = y_mean - slope * x_mean;

    Anim(Line { y_intercept, slope })
}

#[cfg(all(test, feature = "alloc"))]
mod tests {
    use assert_approx_eq::assert_approx_eq;
    extern crate alloc;
    use alloc::vec;

    use super::simple_linear_regression;

    #[test]
    fn test_perfect_regression() {
        let straight_line = vec![(1.0, 1.0), (2.0, 2.0)];
        let line = simple_linear_regression(straight_line.as_slice());
        assert_approx_eq!(line.eval(1.0), 1.0f64);
        assert_approx_eq!(line.eval(10.0), 10.0);
        assert_approx_eq!(line.eval(-10.0), -10.0);
    }

    #[test]
    fn test_negative_perfect_regression() {
        let straight_line = vec![(1.0, -1.0), (2.0, -2.0)];
        let line = simple_linear_regression(straight_line.as_slice());
        assert_approx_eq!(line.eval(1.0), -1.0f64);
        assert_approx_eq!(line.eval(10.0), -10.0);
        assert_approx_eq!(line.eval(-10.0), 10.0);
    }
}


src/anim_box.rs
use std::ops::{Deref, Sub};

use crate::{Anim, Fun};

pub type AnimBox<T, V> = Anim<Box<dyn Fun<T = T, V = V>>>;

impl<F> Anim<F>
where
    F: Fun + 'static,
{
    /// Returns a boxed version of this animation.
    ///
    /// This may be used to reduce the compilation time of deeply nested
    /// animations.
    pub fn into_box(self) -> AnimBox<F::T, F::V> {
        Anim(Box::new(self.0))
    }

    pub fn into_box_fn(self) -> Box<dyn Fn(F::T) -> F::V> {
        Box::new(self.into_fn())
    }
}

// TODO: We need to get rid of the 'static requirements.
impl<F> Anim<F>
where
    F: Fun + 'static,
    F::T: Copy + PartialOrd + Sub<Output = F::T> + 'static,
    F::V: 'static,
{
    pub fn seq_box<G, A>(self, self_end: F::T, next: A) -> AnimBox<F::T, F::V>
    where
        G: Fun<T = F::T, V = F::V> + 'static,
        A: Into<Anim<G>>,
    {
        self.into_box()
            .seq(self_end, next.into().into_box())
            .into_box()
    }
}

impl<'a, T, V> Fun for Box<dyn Fun<T = T, V = V>> {
    type T = T;
    type V = V;

    fn eval(&self, t: Self::T) -> Self::V {
        self.deref().eval(t)
    }
}


examples/plots.rs
use gnuplot::{AutoOption, AxesCommon, Color, Figure, LineWidth};

use pareen::{Anim, Fun};

fn main() {
    let mut plots = Plots { plots: Vec::new() };

    plots.add("id", pareen::id());
    plots.add("lerp between 2 and 4", pareen::lerp(2.0, 4.0));
    plots.add(
        "dynamic lerp between sin^2 and cos",
        pareen::circle().sin().powi(2).lerp(pareen::circle().cos()),
    );
    plots.add(
        "dynamic lerp, squeezed into [0.5 .. 1]",
        pareen::circle()
            .sin()
            .powi(2)
            .lerp(pareen::circle().cos())
            .squeeze_and_surround(0.5..=1.0, 0.0),
    );
    plots.add(
        "switch from 1 to 2 at time=0.5",
        pareen::c(1.0).switch(0.5, 2.0),
    );

    #[cfg(feature = "easer")]
    plots.add(
        "ease transition from 2 to a proportional anim",
        pareen::constant(2.0).seq_ease_in_out(
            0.5,
            easer::functions::Cubic,
            0.3,
            pareen::prop(1.0f32),
        ),
    );

    plots.show_gnuplot();
}

fn sample(n: usize, max_t: f32, anim: Anim<impl Fun<T = f32, V = f32>>) -> (Vec<f32>, Vec<f32>) {
    let mut ts = Vec::new();
    let mut vs = Vec::new();

    for i in 0..=n {
        let time = i as f32 / n as f32 * max_t;
        let value = anim.eval(time);

        ts.push(time);
        vs.push(value);
    }

    (ts, vs)
}

struct Plot {
    name: &'static str,
    ts: Vec<f32>,
    vs: Vec<f32>,
}

struct Plots {
    plots: Vec<Plot>,
}

impl Plots {
    fn add(&mut self, name: &'static str, anim: pareen::Anim<impl pareen::Fun<T = f32, V = f32>>) {
        let (ts, vs) = sample(1000, 1.0, anim);

        self.plots.push(Plot { name, ts, vs });
    }

    fn show_gnuplot(&self) {
        let mut figure = Figure::new();

        // Show plots in a square rows/columns layout
        let n_cols = (self.plots.len() as f32).sqrt() as u32;
        let n_rows = (self.plots.len() as f32).sqrt().ceil() as u32;

        for (i, plot) in self.plots.iter().enumerate() {
            figure
                .axes2d()
                .lines(&plot.ts, &plot.vs, &[Color("blue"), LineWidth(3.0)])
                .set_title(&plot.name, &[])
                .set_x_label("time", &[])
                .set_y_label("value", &[])
                .set_x_ticks(Some((AutoOption::Fix(0.5), 0)), &[], &[])
                .set_y_ticks(Some((AutoOption::Fix(0.5), 0)), &[], &[])
                .set_pos_grid(n_rows, n_cols, i as u32);
        }

        figure.show().unwrap();
    }
}


src/arithmetic.rs
use std::ops::{Add, Mul, Neg, Sub};

use crate::{primitives::ConstantClosure, Anim, Fun};

impl<F, G> Add<Anim<G>> for Anim<F>
where
    F: Fun,
    G: Fun<T = F::T>,
    F::V: Add<G::V>,
{
    type Output = Anim<AddClosure<F, G>>;

    fn add(self, rhs: Anim<G>) -> Self::Output {
        Anim(AddClosure(self.0, rhs.0))
    }
}

impl<W, F> Add<W> for Anim<F>
where
    W: Copy,
    F: Fun,
    F::V: Add<W>,
{
    type Output = Anim<AddClosure<F, ConstantClosure<F::T, W>>>;

    fn add(self, rhs: W) -> Self::Output {
        Anim(AddClosure(self.0, ConstantClosure::from(rhs)))
    }
}

impl<F, G> Sub<Anim<G>> for Anim<F>
where
    F: Fun,
    G: Fun<T = F::T>,
    F::V: Sub<G::V>,
{
    type Output = Anim<SubClosure<F, G>>;

    fn sub(self, rhs: Anim<G>) -> Self::Output {
        Anim(SubClosure(self.0, rhs.0))
    }
}

impl<W, F> Sub<W> for Anim<F>
where
    W: Copy,
    F: Fun,
    F::T: Copy,
    F::V: Sub<W>,
{
    type Output = Anim<SubClosure<F, ConstantClosure<F::T, W>>>;

    fn sub(self, rhs: W) -> Self::Output {
        Anim(SubClosure(self.0, ConstantClosure::from(rhs)))
    }
}

impl<F, G> Mul<Anim<G>> for Anim<F>
where
    F: Fun,
    G: Fun<T = F::T>,
    F::T: Copy,
    F::V: Mul<G::V>,
{
    type Output = Anim<MulClosure<F, G>>;

    fn mul(self, rhs: Anim<G>) -> Self::Output {
        Anim(MulClosure(self.0, rhs.0))
    }
}

impl<W, F> Mul<W> for Anim<F>
where
    W: Copy,
    F: Fun,
    F::T: Copy,
    F::V: Mul<W>,
{
    type Output = Anim<MulClosure<F, ConstantClosure<F::T, W>>>;

    fn mul(self, rhs: W) -> Self::Output {
        Anim(MulClosure(self.0, ConstantClosure::from(rhs)))
    }
}

impl<V, F> Neg for Anim<F>
where
    V: Copy,
    F: Fun<V = V>,
{
    type Output = Anim<NegClosure<F>>;

    fn neg(self) -> Self::Output {
        Anim(NegClosure(self.0))
    }
}

#[doc(hidden)]
#[derive(Debug, Clone)]
pub struct AddClosure<F, G>(F, G);

impl<F, G> Fun for AddClosure<F, G>
where
    F: Fun,
    G: Fun<T = F::T>,
    F::T: Clone,
    F::V: Add<G::V>,
{
    type T = F::T;
    type V = <F::V as Add<G::V>>::Output;

    fn eval(&self, t: F::T) -> Self::V {
        self.0.eval(t.clone()) + self.1.eval(t)
    }
}

#[doc(hidden)]
#[derive(Debug, Clone)]
pub struct SubClosure<F, G>(F, G);

impl<F, G> Fun for SubClosure<F, G>
where
    F: Fun,
    G: Fun<T = F::T>,
    F::T: Clone,
    F::V: Sub<G::V>,
{
    type T = F::T;
    type V = <F::V as Sub<G::V>>::Output;

    fn eval(&self, t: F::T) -> Self::V {
        self.0.eval(t.clone()) - self.1.eval(t)
    }
}

#[doc(hidden)]
#[derive(Debug, Clone)]
pub struct MulClosure<F, G>(F, G);

impl<F, G> Fun for MulClosure<F, G>
where
    F: Fun,
    G: Fun<T = F::T>,
    F::T: Copy,
    F::V: Mul<G::V>,
{
    type T = F::T;
    type V = <F::V as Mul<G::V>>::Output;

    fn eval(&self, t: F::T) -> Self::V {
        self.0.eval(t) * self.1.eval(t)
    }
}

#[doc(hidden)]
pub struct NegClosure<F>(F);

impl<F> Fun for NegClosure<F>
where
    F: Fun,
    F::V: Neg,
{
    type T = F::T;
    type V = <F::V as Neg>::Output;

    fn eval(&self, t: F::T) -> Self::V {
        -self.0.eval(t)
    }
}


src/primitives.rs
use core::marker::PhantomData;
use core::ops::Mul;

use crate::{Anim, Fun};

use num_traits::{Float, FloatConst};

/// Turn any function `Fn(T) -> V` into an [`Anim`](struct.Anim.html).
///
/// # Example
/// ```
/// # use assert_approx_eq::assert_approx_eq;
/// fn my_crazy_function(t: f32) -> f32 {
///     42.0 / t
/// }
///
/// let anim = pareen::fun(my_crazy_function);
///
/// assert_approx_eq!(anim.eval(1.0), 42.0);
/// assert_approx_eq!(anim.eval(2.0), 21.0);
/// ```
pub fn fun<T, V>(f: impl Fn(T) -> V) -> Anim<impl Fun<T = T, V = V>> {
    From::from(f)
}

struct WrapFn<T, V, F: Fn(T) -> V>(F, PhantomData<(T, V)>);

impl<T, V, F> From<F> for Anim<WrapFn<T, V, F>>
where
    F: Fn(T) -> V,
{
    fn from(f: F) -> Self {
        Anim(WrapFn(f, PhantomData))
    }
}

impl<T, V, F> Fun for WrapFn<T, V, F>
where
    F: Fn(T) -> V,
{
    type T = T;
    type V = V;

    fn eval(&self, t: T) -> V {
        self.0(t)
    }
}

/// A constant animation, always returning the same value.
///
/// # Example
/// ```
/// # use assert_approx_eq::assert_approx_eq;
/// let anim = pareen::constant(1.0f32);
///
/// assert_approx_eq!(anim.eval(-10000.0f32), 1.0);
/// assert_approx_eq!(anim.eval(0.0), 1.0);
/// assert_approx_eq!(anim.eval(42.0), 1.0);
/// ```
pub fn constant<T, V: Clone>(c: V) -> Anim<impl Fun<T = T, V = V>> {
    fun(move |_| c.clone())
}

#[doc(hidden)]
#[derive(Debug, Clone)]
pub struct ConstantClosure<T, V>(V, PhantomData<T>);

impl<T, V> Fun for ConstantClosure<T, V>
where
    V: Clone,
{
    type T = T;
    type V = V;

    fn eval(&self, _: T) -> V {
        self.0.clone()
    }
}

impl<T, V> From<V> for ConstantClosure<T, V>
where
    V: Clone,
{
    fn from(v: V) -> Self {
        ConstantClosure(v, PhantomData)
    }
}

impl<T, V> From<V> for Anim<ConstantClosure<T, V>>
where
    V: Clone,
{
    fn from(v: V) -> Self {
        Anim(ConstantClosure::from(v))
    }
}

/// An animation that returns a value proportional to time.
///
/// # Example
///
/// Scale time with a factor of three:
/// ```
/// # use assert_approx_eq::assert_approx_eq;
/// let anim = pareen::prop(3.0f32);
///
/// assert_approx_eq!(anim.eval(0.0f32), 0.0);
/// assert_approx_eq!(anim.eval(3.0), 9.0);
/// ```
pub fn prop<T, V, W>(m: V) -> Anim<impl Fun<T = T, V = W>>
where
    V: Clone + Mul<Output = W> + From<T>,
{
    fun(move |t| m.clone() * From::from(t))
}

/// An animation that returns time as its value.
///
/// This is the same as [`prop`](fn.prop.html) with a factor of one.
///
/// # Examples
/// ```
/// let anim = pareen::id::<isize, isize>();
///
/// assert_eq!(anim.eval(-100), -100);
/// assert_eq!(anim.eval(0), 0);
/// assert_eq!(anim.eval(100), 100);
/// ```
/// ```
/// # use assert_approx_eq::assert_approx_eq;
/// let anim = pareen::id::<f32, f32>() * 3.0 + 4.0;
///
/// assert_approx_eq!(anim.eval(0.0), 4.0);
/// assert_approx_eq!(anim.eval(100.0), 304.0);
/// ```
pub fn id<T, V>() -> Anim<impl Fun<T = T, V = V>>
where
    V: From<T>,
{
    fun(From::from)
}

/// Proportionally increase value from zero to 2π.
pub fn circle<T, V>() -> Anim<impl Fun<T = T, V = V>>
where
    T: Float,
    V: Float + FloatConst + From<T>,
{
    prop(V::PI() * (V::one() + V::one()))
}

/// Proportionally increase value from zero to π.
pub fn half_circle<T, V>() -> Anim<impl Fun<T = T, V = V>>
where
    T: Float,
    V: Float + FloatConst + From<T>,
{
    prop(V::PI())
}

/// Proportionally increase value from zero to π/2.
pub fn quarter_circle<T, V>() -> Anim<impl Fun<T = T, V = V>>
where
    T: Float,
    V: Float + FloatConst + From<T>,
{
    prop(V::PI() * (V::one() / (V::one() + V::one())))
}

/// Evaluate a quadratic polynomial in time.
pub fn quadratic<T>(w: &[T; 3]) -> Anim<impl Fun<T = T, V = T> + '_>
where
    T: Float,
{
    fun(move |t| {
        let t2 = t * t;

        w[0] * t2 + w[1] * t + w[2]
    })
}

/// Evaluate a cubic polynomial in time.
pub fn cubic<T>(w: &[T; 4]) -> Anim<impl Fun<T = T, V = T> + '_>
where
    T: Float,
{
    fun(move |t| {
        let t2 = t * t;
        let t3 = t2 * t;

        w[0] * t3 + w[1] * t2 + w[2] * t + w[3]
    })
}

/// Count from 0 to `end` (non-inclusive) cyclically, at the given frames per
/// second rate.
///
/// # Example
/// ```
/// let anim = pareen::cycle(3, 5.0);
/// assert_eq!(anim.eval(0.0), 0);
/// assert_eq!(anim.eval(0.1), 0);
/// assert_eq!(anim.eval(0.3), 1);
/// assert_eq!(anim.eval(0.5), 2);
/// assert_eq!(anim.eval(0.65), 0);
///
/// assert_eq!(anim.eval(-0.1), 2);
/// assert_eq!(anim.eval(-0.3), 1);
/// assert_eq!(anim.eval(-0.5), 0);
/// ```
pub fn cycle(end: usize, fps: f32) -> Anim<impl Fun<T = f32, V = usize>> {
    fun(move |t: f32| {
        if t < 0.0 {
            let tau = (t.abs() * fps) as usize;

            end - 1 - tau % end
        } else {
            let tau = (t * fps) as usize;

            tau % end
        }
    })
}


src/anim_with_dur.rs
use core::ops::{Add, Div, Mul, Sub};

use num_traits::{Float, One};

use crate::{Anim, Fun};

/// An `Anim` together with the duration that it is intended to be played for.
///
/// Explicitly carrying the duration around makes it easier to sequentially
/// compose animations in some places.
#[derive(Clone, Debug)]
pub struct AnimWithDur<F: Fun>(pub Anim<F>, pub F::T);

impl<F> Anim<F>
where
    F: Fun,
{
    /// Tag this animation with the duration that it is intended to be played
    /// for.
    ///
    /// Note that using this tagging is completely optional, but it may
    /// make it easier to combine animations sometimes.
    pub fn dur(self, t: F::T) -> AnimWithDur<F> {
        AnimWithDur(self, t)
    }
}

impl<'a, V> From<&'a [V]> for AnimWithDur<SliceClosure<'a, V>>
where
    V: Clone,
{
    fn from(slice: &'a [V]) -> Self {
        AnimWithDur(Anim(SliceClosure(slice)), slice.len())
    }
}

pub fn slice<'a, V>(slice: &'a [V]) -> AnimWithDur<impl Fun<T = usize, V = V> + 'a>
where
    V: Clone + 'a,
{
    slice.into()
}

#[doc(hidden)]
pub struct SliceClosure<'a, V>(&'a [V]);

impl<'a, V> Fun for SliceClosure<'a, V>
where
    V: Clone,
{
    type T = usize;
    type V = V;

    fn eval(&self, t: Self::T) -> Self::V {
        self.0[t].clone()
    }
}

impl<F> Anim<F>
where
    F: Fun,
    F::T: Clone + Float,
{
    pub fn scale_to_dur(self, dur: F::T) -> AnimWithDur<impl Fun<T = F::T, V = F::V>> {
        self.scale_time(F::T::one() / dur).dur(dur)
    }
}

impl<F> AnimWithDur<F>
where
    F: Fun,
    F::T: Clone,
{
    pub fn as_ref(&self) -> AnimWithDur<&F> {
        AnimWithDur(self.0.as_ref(), self.1.clone())
    }
}

impl<F> AnimWithDur<F>
where
    F: Fun,
{
    pub fn transform<G, H>(self, h: H) -> AnimWithDur<G>
    where
        G: Fun<T = F::T>,
        H: FnOnce(Anim<F>) -> Anim<G>,
    {
        AnimWithDur(h(self.0), self.1)
    }

    pub fn map<W>(self, f: impl Fn(F::V) -> W) -> AnimWithDur<impl Fun<T = F::T, V = W>> {
        self.transform(move |anim| anim.map(f))
    }

    pub fn dur(self, t: F::T) -> AnimWithDur<F> {
        AnimWithDur(self.0, t)
    }
}

impl<'a, T, X, Y, F> AnimWithDur<F>
where
    T: 'a + Clone,
    X: 'a,
    Y: 'a,
    F: Fun<T = T, V = (X, Y)>,
{
    pub fn unzip(
        &'a self,
    ) -> (
        AnimWithDur<impl Fun<T = F::T, V = X> + 'a>,
        AnimWithDur<impl Fun<T = F::T, V = Y> + 'a>,
    ) {
        (
            self.as_ref().transform(|anim| anim.fst()),
            self.as_ref().transform(|anim| anim.snd()),
        )
    }
}

impl<F> AnimWithDur<F>
where
    F: Fun,
    F::T: Copy + PartialOrd + Sub<Output = F::T>,
{
    pub fn seq<G>(self, next: Anim<G>) -> Anim<impl Fun<T = F::T, V = F::V>>
    where
        G: Fun<T = F::T, V = F::V>,
    {
        self.0.seq(self.1, next)
    }
}

impl<F> AnimWithDur<F>
where
    F: Fun,
    F::T: Copy + PartialOrd + Sub<Output = F::T> + Add<Output = F::T>,
{
    pub fn seq_with_dur<G>(self, next: AnimWithDur<G>) -> AnimWithDur<impl Fun<T = F::T, V = F::V>>
    where
        G: Fun<T = F::T, V = F::V>,
    {
        let dur = self.1.clone() + next.1;
        AnimWithDur(self.seq(next.0), dur)
    }
}

impl<F> AnimWithDur<F>
where
    F: Fun,
    F::T: Clone + Float,
{
    pub fn repeat(self) -> Anim<impl Fun<T = F::T, V = F::V>> {
        self.0.repeat(self.1)
    }
}

impl<F> AnimWithDur<F>
where
    F: Fun,
    F::T: Clone + Sub<Output = F::T>,
{
    pub fn backwards(self) -> AnimWithDur<impl Fun<T = F::T, V = F::V>> {
        AnimWithDur(self.0.backwards(self.1.clone()), self.1)
    }
}

impl<F> AnimWithDur<F>
where
    F: Fun,
    F::T: Clone + Mul<Output = F::T> + Div<Output = F::T>,
{
    pub fn scale_time(self, t_scale: F::T) -> AnimWithDur<impl Fun<T = F::T, V = F::V>> {
        AnimWithDur(self.0.scale_time(t_scale.clone()), self.1 / t_scale)
    }
}

#[macro_export]
macro_rules! seq_with_dur {
    (
        $expr:expr $(,)?
    ) => {
        $expr
    };

    (
        $head:expr,
        $($tail:expr $(,)?)+
    ) => {
        $head.seq_with_dur($crate::seq_with_dur!($($tail,)*))
    }
}


src/lib.rs
//! Pareen is a small library for *par*ameterized inbetw*een*ing.
//! The intended application is in game programming, where you sometimes have
//! two discrete game states between which you want to transition smoothly
//! for visualization purposes.
//!
//! Pareen gives you tools for composing animations that are parameterized by
//! time (i.e. mappings from time to some animated value) without constantly
//! having to pass around time variables; it hides the plumbing, so that you
//! need to provide time only once: when evaluating the animation.
//!
//! Animations are composed similarly to Rust's iterators, so no memory
//! allocations are necessary.
//! ## Examples
//!
//! ```rust
//! # use assert_approx_eq::assert_approx_eq;
//! // An animation returning a constant value
//! let anim1 = pareen::constant(1.0f64);
//!
//! // Animations can be evaluated at any time
//! let value = anim1.eval(0.5);
//!
//! // Animations can be played in sequence
//! let anim2 = anim1.seq(0.7, pareen::prop(0.25) + 0.5);
//!
//! // Animations can be composed and transformed in various ways
//! let anim3 = anim2
//!     .lerp(pareen::circle().cos())
//!     .scale_min_max(5.0, 10.0)
//!     .backwards(1.0)
//!     .squeeze(0.5..=1.0);
//!
//! let anim4 = pareen::cubic(&[1.0, 2.0, 3.0, 4.0]) - anim3;
//!
//! let value = anim4.eval(1.0);
//! assert_approx_eq!(value, 0.0);
//! ```

#![no_std]

mod anim;
#[cfg(feature = "alloc")]
mod anim_box;
mod anim_with_dur;
mod arithmetic;
mod primitives;

pub mod stats;

#[cfg(feature = "easer")]
mod easer_combinators;

pub use anim::{cond, lerp, Anim, Fun};
#[cfg(feature = "alloc")]
pub use anim_box::AnimBox;
pub use anim_with_dur::{slice, AnimWithDur};
pub use primitives::{
    circle, constant, cubic, cycle, fun, half_circle, id, prop, quadratic, quarter_circle,
};
pub use stats::{simple_linear_regression, simple_linear_regression_with_slope};

#[cfg(feature = "easer")]
pub use easer;

#[cfg(feature = "easer")]
pub use easer_combinators::{ease_in, ease_in_out, ease_out};
